clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/client/P2PFileTransferClientHandler.java

```diff
@@ -261,16 +270,62 @@ private String getFileNameFromHeader(HttpResponse response) {
 
   private void handleEndOfTransfer(ChannelHandlerContext ctx) {
     LOGGER.info("All files received successfully for {}", payload.getFullResourceName());
-    // In the short term, we decided to let the netty client handle writing files to the disk, so
-    // the future completes with a null. It's subject to change.
-    inputStreamFuture.toCompletableFuture().complete(null);
-    ctx.close();
+
+    try {
+      RocksDBUtils.renameTempTransferredPartitionDirToPartitionDir(
+          payload.getBaseDir(),
+          payload.getTopicName(),
+          payload.getPartition());
+      LOGGER.info(
+          "Renamed temp partition dir to partition dir for {}",
+          Utils.getReplicaId(payload.getTopicName(), payload.getPartition()));
+      inputStreamFuture.toCompletableFuture().complete(null);
+    } catch (Exception e) {
+      LOGGER.error(
+          "Failed to rename temp partition dir to partition dir for {}. Even all the files are received, "
+              + "the transfer future will be completed exceptionally. ",
+          Utils.getReplicaId(payload.getTopicName(), payload.getPartition()),
+          e);
+      // Complete future exceptionally, no need to do resource cleanup here because if files are all received, means
```

Comment:
Could you explain a bit more here? If rename would fail (which I cannot think of why given you already handle the folder check before transfer), then shouldn't we invoke the cleanupresources() method below?

Reply:
Synced offline, now it makes sense to me as the change is also captured in handlePeerFetchException